actionpack/lib/action_dispatch/middleware/stack.rb

```diff
@@ -176,12 +187,23 @@ def assert_index(index, where)
       end
 
       def build_middleware(klass, args, block)
+        @rack_sendfile_deprecated = false if klass == Rack::Sendfile
         Middleware.new(klass, args, block)
       end
 
       def index_of(klass)
+        raise "ActionDispatch::MiddlewareStack::FakeSendfile can not be referenced in middleware operations" if klass == FakeSendfile
+
+        if klass == Rack::Sendfile && @rack_sendfile_deprecated
+          ActiveSupport::Deprecation.warn(<<-MSG.squish)
+            Rack::Sendfile is removed from the default middleware stack in Rails
+            and referencing it in middleware operations without adding it back
+            is deprecated and will throw an error in future versions of Rails.
+          MSG
+        end
+
         middlewares.index do |m|
-          m.name == klass.name
+          m.name == klass.name || (@rack_sendfile_deprecated && m.klass == FakeSendfile && klass == Rack::Sendfile)
```

Comment:
Is there another way to implement this? I'm not sure we need to be changing this class for this.

Does this message only occur when using `insert_before` or `insert_after`? That feels like too narrow of a scope to me. 🤔

Reply:
@zzak, Thank you for the reply.

> Is there another way to implement this? I'm not sure we need to be changing this class for this.

I think this is a reasonable place for the change, since we can easily check if the `Rack::Sendfile` is explicitly added and to also plug to every stack referencing method.
Do you have anything particular in mind for an alternative implementation?

> Does this message only occur when using `insert_before` or `insert_after`? That feels like too narrow of a scope to me.

This is valid for any middleware referencing operations, not limited to `insert_before` or `insert_after`.
Here is the full list - `insert`/`insert_before`, `insert_after`, `swap`, `move`/`move_before`, `move_after`

The idea is to be backwards compatible with Rails projects, which depend on the `Rack::Sendfile` middleware to be present by default, and is referenced in example for `insert_after`. However we do not want to raise deprecation warnings if the `Rack::Sendfile` is explicitly added in the stack(e.g. `use`), since it would still be present after we've removed it from the default middleware stack in Rails.
All of this would of course be removed completely after some time/version update.

Reply:
> Do you have anything particular in mind for an alternative implementation?

I think we should consider an alternative, but I can't think of what that is atm..

> This is valid for any middleware referencing operation, not limited to insert_before or insert_after.
> Here is the full list - insert/insert_before, insert_after, swap, move/move_before, move_after

In this case, how do you feel about adding test cases for those scenarios? Assuming the behavior remains the same even if we refactor the implementation.

Reply:
On second thought, I wonder if we **even** need to go so far as to issue a warning when using `insert_before`, etc.

If we simply want to remove it from the default stack, just replace it with a fake one, that fake middleware should issue the warning and then perform the typical SendFile functions until we can remove it.

If a user wants to intentionally use SendFile, then that is on them.

Maybe there is a way to monkey-patch SendFile to emit the warning for us, but I think that the warning should come from there (when the user intentionally is using SendFile). Unlike, if they are using it based on it being available by default. Does that make sense?

Reply:
> On second thought, I wonder if we even need to go so far as to issue a warning when using insert_before, etc.

> If we simply want to remove it from the default stack, just replace it with a fake one, that fake middleware should issue the warning and then perform the typical SendFile functions until we can remove it.

If I understand correctly, the proposal is to keep `Sendfile` functions and only issue a warning?
Currently, this is implemented to actually remove the `Rack::Sendfile` from the default middleware stack, because of this direction:
https://github.com/rails/rails/issues/41148#issuecomment-1007877623
> We should remove all configs that are used by `Rack::Sendfile`. Rails should not include that in the middleware stack anymore. If uses want to use `Rack::Sendfile`, and they should be able to, they will have to explicit call use `Rack::SendFile, 'X-Sendfile'`.

I just thought it would be nice to not introduce a breaking change in the middleware operations, which use the `Rack::Sendfile` as a target.

If we want to actually keep the `Rack::SendFile` in the stack and only issue a warning, then maybe we need to completely scrap this PR. Yet I will still add the additionally mentioned test cases, just in case.

> In this case, how do you feel about adding test cases for those scenarios? Assuming the behavior remains the same even if we refactor the implementation.

Reply:
I think there are 2 possible scenarios here:

1. As I just re-read the original thread, there is potential to make Rails version of SendFile safer, therefor keeping it in
2. We remove it like your PR here, and RF initially suggested

Assuming that this PR is going to focus on point 2, my question is mainly on how we communicate that breaking change to users. Which I assume is something we want to do in that case.

So if we're going to remove functionality without a deprecation path, what is the best way to communicate that here?

I think we're not on the same page about what exactly we want to do here, so let's figure that out.

For example, should we really remove `config.action_dispatch.x_sendfile_header`? Is the default for this value enabled? Or can we just make the new default to false?

Also, sorry for all of the back and forth I may be making this more complicated that it truly is.

Reply:
> For example, should we really remove config.action_dispatch.x_sendfile_header? Is the default for this value enabled? Or can we just make the new default to false?

I've brought this up in the discussion https://github.com/rails/rails/issues/41148#issuecomment-1007819773
And the response https://github.com/rails/rails/issues/41148#issuecomment-1007877623 stated that we should remove all configs that are used by `Rack::Sendfile`. As I've already said, that's why this PR is in that direction.

I've added the additional test cases which we discussed.

I'm not sure if and how we can communicate this change better. What do you think?

Feel free to close this PR if this is not your desired implementation and would like to focus on the another approach.
> 1. As I just re-read the original thread, there is potential to make Rails version of SendFile safer, therefor keeping it in

Reply:
Let's resolve this thread so others can review this PR as it is.

Thanks for clearing up my questions!